### ISSUE TITLE
IOException: file is too large

### ISSUE THREAD

**Issue by nnovikov** [upx/upx#374]
What is the maximum file size supported by the upx?
I try to compress 1G elf on linux and have error:
`upx: foobar: IOException: file is too large -- skipped`

upx 3.95
Debian GNU/Linux 10 (buster)
Intel(R) Core(TM) i7-8550U CPU

**Comment by jreiser** [comment]
conf.h:#define UPX_RSIZE_MAX_MEM   (768 * 1024 * 1024)   // DO NOT CHANGE
because (2+epsilon)*UPX_RSIZE_MAX_MEM must fit on a 32-bit machine.

**Comment by jreiser** [comment]
An experiment a few years ago showed that somewhere around 900MiB was the actual limit on a 32-bit Linux system.
Possible workaround: remove everything but actual code+data; namely: debugging info, etc.

**Comment by jreiser** [comment]
Modifying UPX_RSIZE_MAX_MEM is open for experiment, just like anything.  But the run-time decompression stubs have some hard limits of 4GiB (and possibly 2GiB).  For example, `grep XXX stub/src/amd64-linux.elf-entry.S`.

**Comment by nnovikov** [comment]
Ok, big thanks for explanation.

**Comment by lostmsu** [comment]
Just faced this trying to compress `_pywrap_tensorflow_internal.pyd` of some 1.3GB on 64-bit Windows.

**Comment by markus-oberhumer** [comment]
We cannot increase UPX_RSIZE_MAX_MEM because that will cause integer overflows.

But why are your executable so large in the first place?

1) if you have debug info, then strip the binaries before running upx

2) if you have large data sections or overlays, then append or attach this data after running upx

**Comment by jojje** [comment]
That is a question for nVidia and Facebook. CUDA and tensorflow libraries are hundreds of megabyte, to over a gigabyte in size. CUDA is a proprietary blob, and as such your suggestions above don't help for this particular case.

Example, `libtorch_cuda.so` in tensorflow 2.0.1 is `1075112584` bytes. And that's just _one_ of the many gigantic files in that popular framework.

Would be fantastic if UPX could be enhanced to cope with the crazy multi-gigabyte library world of ML.

**Comment by markus-oberhumer** [comment]
WONTFIX

CUDA should handle this like other programs that deal with large binary blobs/assets (i.e. games) and handle compression internally.

**Comment by jreiser** [comment]
`Example, libtorch_cuda.so in tensorflow 2.0.1 is 1075112584 bytes. And that's just one of the many gigantic files in that popular framework.`
How compressible is libtorch_cuda.so and others?  If it's (mostly) already compressed (by any method at all) then the possible savings from compression by UPX would be small.  Compress libtorch_cuda.so using zip, gzip, 7zip, rar, anything at all.  How much space savings do you see?  [Numerical data, please.] That would give an approximate answer to how much UPX compression might save, if UPX had no size limit.

The simplifications and efficiencies of being able to have 2 copies of the file in the address space at the same time are large.

**Comment by jreiser** [comment]
If `libtorch_cuda.so` is a real ELF ET_DYN shared library, then please run `readelf --headers libtorch_cuda.so` and post the output here. I'd like to see the data on the [PT_}LOAD segments, the Sections, etc.

Here's how to get experience with a compressed game or other big app.  Put all the required files into a directory that is a subtree laid out like the app expects.  Build a `squashfs` whose root is that directory.  The result is an ordinary file that is an image of the squashed subtree.  Inside the image, blocks of 128KiB have been independently compressed using zlib, and this reduces space.  The image can be mounted using a loop-back mount, which makes the tree visible just like a ordinary tree, except that writes won't work.  Also, the squashfs image can be written to a USB flash drive (using 'cp` or 'dd' to the whole drive [no partitions]), and then the flash drive can be mounted like any other filesystem.  Blocks of the squashfs will be de-compressed on demand; the de-compressed blocks are cached into RAM and then mmap()ed as regular blocks.

**Comment by tklausmann** [comment]
This isn't CUDA but the equivalent for AMD GPUs (HIP). The shared object file is about 1.3G (1'351'048'265 bytes) in its base form. gzip(1) with default params crunches it to 259M (271'514'680 bytes). So there is definitely compression potential in there. While the squashfs approach is something we (as in my team) might look into, it's a bit less transparent than using upx would be --- though it would have the advantage of guaranteeing that the .so files still work. I've had proprietary binaries in the past that would do _something_ that breaks when upx'd.

```
$ /readelf --headers  /opt/lib/python/site-packages/torch/lib/libtorch_hip.so 
ELF Header:
  Magic:   7f 45 4c 46 02 01 01 03 00 00 00 00 00 00 00 00 
  Class:                             ELF64
  Data:                              2's complement, little endian
  Version:                           1 (current)
  OS/ABI:                            UNIX - GNU
  ABI Version:                       0
  Type:                              DYN (Shared object file)
  Machine:                           Advanced Micro Devices X86-64
  Version:                           0x1
  Entry point address:               0x689f60
  Start of program headers:          64 (bytes into file)
  Start of section headers:          1307933992 (bytes into file)
  Flags:                             0x0
  Size of this header:               64 (bytes)
  Size of program headers:           56 (bytes)
  Number of program headers:         21
  Size of section headers:           64 (bytes)
  Number of section headers:         32
  Section header string table index: 28

Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .dynsym           DYNSYM           00000000000355b8  000355b8
       00000000000c5868  0000000000000018   A      31     1     8
  [ 2] .gnu.version      VERSYM           00000000004b29ca  004b29ca
       000000000001075e  0000000000000002   A       1     0     2
  [ 3] .gnu.version_r    VERNEED          00000000004c3128  004c3128
       0000000000000270  0000000000000000   A      31    10     8
  [ 4] .rela.dyn         RELA             00000000004c3398  004c3398
       000000000015fc60  0000000000000018   A       1     0     8
  [ 5] .rela.plt         RELA             0000000000622ff8  00622ff8
       000000000003d6c8  0000000000000018  AI       1    21     8
  [ 6] .init             PROGBITS         0000000000661000  00661000
       000000000000001a  0000000000000000  AX       0     0     4
  [ 7] .plt              PROGBITS         0000000000661020  00661020
       0000000000028f40  0000000000000010  AX       0     0     16
  [ 8] .text             PROGBITS         0000000000689f60  00689f60
       0000000001afbc0a  0000000000000000  AX       0     0     16
  [ 9] .fini             PROGBITS         0000000002185b6c  02185b6c
       0000000000000009  0000000000000000  AX       0     0     4
  [10] .rodata           PROGBITS         0000000002186000  02186000
       000000000a2d44e0  0000000000000000   A       0     0     32
  [11] .hip_fatbin       PROGBITS         000000000c45b000  0c45b000
       000000004028d420  0000000000000000   A       0     0     4096
  [12] .eh_frame_hdr     PROGBITS         000000004c6e8420  4c6e8420
       0000000000072694  0000000000000000   A       0     0     4
  [13] .eh_frame         PROGBITS         000000004c75aab8  4c75aab8
       000000000038c7b0  0000000000000000   A       0     0     8
  [14] .gcc_except_table PROGBITS         000000004cae7268  4cae7268
       0000000000116db0  0000000000000000   A       0     0     4
  [15] .tbss             NOBITS           000000004cbff848  4cbfe848
       0000000000000058  0000000000000000 WAT       0     0     8
  [16] .init_array       INIT_ARRAY       000000004cbff848  4cbfe848
       0000000000002390  0000000000000008  WA       0     0     8
  [17] .fini_array       FINI_ARRAY       000000004cc01bd8  4cc00bd8
       0000000000000008  0000000000000008  WA       0     0     8
  [18] .data.rel.ro      PROGBITS         000000004cc01be0  4cc00be0
       0000000000061378  0000000000000000  WA       0     0     32
  [19] .dynamic          DYNAMIC          000000004cc62f58  4cc61f58
       0000000000000330  0000000000000010  WA      31     0     8
  [20] .got              PROGBITS         000000004cc63288  4cc62288
       0000000000016d60  0000000000000008  WA       0     0     8
  [21] .got.plt          PROGBITS         000000004cc7a000  4cc79000
       00000000000147b0  0000000000000008  WA       0     0     8
  [22] .data             PROGBITS         000000004cc8e7c0  4cc8d7c0
       0000000000000ae8  0000000000000000  WA       0     0     32
  [23] .hipFatBinSegment PROGBITS         000000004cc8f2a8  4cc8e2a8
       0000000000001260  0000000000000000  WA       0     0     8
  [24] .bss              NOBITS           000000004cc90520  4cc8f508
       000000000003e160  0000000000000000  WA       0     0     32
  [25] .comment          PROGBITS         0000000000000000  4cc8f508
       00000000000000df  0000000000000001  MS       0     0     1
  [26] .symtab           SYMTAB           0000000000000000  4cc8f5e8
       00000000002ac960  0000000000000018          27   83126     8
  [27] .strtab           STRTAB           0000000000000000  4cf3bf48
       000000000101bcaa  0000000000000000           0     0     1
  [28] .shstrtab         STRTAB           0000000000000000  4df57bf2
       0000000000000130  0000000000000000           0     0     1
  [29] .note.gnu.bu[...] NOTE             000000004cccf000  4df59000
       0000000000000024  0000000000000000   A       0     0     4
  [30] .gnu.hash         GNU_HASH         000000004cccf028  4df59028
       000000000003531c  0000000000000000   A       1     0     8
  [31] .dynstr           STRTAB           000000004f234000  504be000
       00000000003b7c44  0000000000000000   A       0     0     8
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  R (retain), D (mbind), l (large), p (processor specific)

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x00000000006606c0 0x00000000006606c0  R      0x1000
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  LOAD           0x0000000000661000 0x0000000000661000 0x0000000000661000
                 0x0000000001b24b75 0x0000000001b24b75  R E    0x1000
  LOAD           0x0000000002186000 0x0000000002186000 0x0000000002186000
                 0x000000004aa78018 0x000000004aa78018  R      0x1000
  GNU_EH_FRAME   0x000000004c6e8420 0x000000004c6e8420 0x000000004c6e8420
                 0x0000000000072694 0x0000000000072694  R      0x4
  LOAD           0x000000004cbfe848 0x000000004cbff848 0x000000004cbff848
                 0x0000000000090cc0 0x00000000000cee38  RW     0x1000
  TLS            0x000000004cbfe848 0x000000004cbff848 0x000000004cbff848
                 0x0000000000000000 0x0000000000000058  R      0x8
  GNU_RELRO      0x000000004cbfe848 0x000000004cbff848 0x000000004cbff848
                 0x000000000007a7b8 0x000000000007a7b8  R      0x1
  DYNAMIC        0x000000004cc61f58 0x000000004cc62f58 0x000000004cc62f58
                 0x0000000000000330 0x0000000000000330  RW     0x8
  NOTE           0x000000004df59000 0x000000004cccf000 0x000000004cccf000
                 0x0000000000000024 0x0000000000000024  R      0x4
  LOAD           0x000000004df59000 0x000000004cccf000 0x000000004cccf000
                 0x00000000003ecf00 0x00000000003ecf00  RW     0x1000
  LOAD           0x000000004e346000 0x000000004d0bc000 0x000000004d0bc000
                 0x00000000003b7bc8 0x00000000003b7bc8  RW     0x1000
  LOAD           0x000000004e6fe000 0x000000004d474000 0x000000004d474000
                 0x00000000003b7bd8 0x00000000003b7bd8  RW     0x1000
  LOAD           0x000000004eab6000 0x000000004d82c000 0x000000004d82c000
                 0x00000000003b7be0 0x00000000003b7be0  RW     0x1000
  LOAD           0x000000004ee6e000 0x000000004dbe4000 0x000000004dbe4000
                 0x00000000003b7bf0 0x00000000003b7bf0  RW     0x1000
  LOAD           0x000000004f226000 0x000000004df9c000 0x000000004df9c000
                 0x00000000003b7c00 0x00000000003b7c00  RW     0x1000
  LOAD           0x000000004f5de000 0x000000004e354000 0x000000004e354000
                 0x00000000003b7c10 0x00000000003b7c10  RW     0x1000
  LOAD           0x000000004f996000 0x000000004e70c000 0x000000004e70c000
                 0x00000000003b7c20 0x00000000003b7c20  RW     0x1000
  LOAD           0x000000004fd4e000 0x000000004eac4000 0x000000004eac4000
                 0x00000000003b7c28 0x00000000003b7c28  RW     0x1000
  LOAD           0x0000000050106000 0x000000004ee7c000 0x000000004ee7c000
                 0x00000000003b7c38 0x00000000003b7c38  RW     0x1000
  LOAD           0x00000000504be000 0x000000004f234000 0x000000004f234000
                 0x00000000003b7c48 0x00000000003b7c48  RW     0x1000

 Section to Segment mapping:
  Segment Sections...
   00     .dynsym .gnu.version .gnu.version_r .rela.dyn .rela.plt 
   01     
   02     .init .plt .text .fini 
   03     .rodata .hip_fatbin .eh_frame_hdr .eh_frame .gcc_except_table 
   04     .eh_frame_hdr 
   05     .init_array .fini_array .data.rel.ro .dynamic .got .got.plt .data .hipFatBinSegment .bss 
   06     .tbss 
   07     .init_array .fini_array .data.rel.ro .dynamic .got 
   08     .dynamic 
   09     .note.gnu.build-id 
   10     .note.gnu.build-id .gnu.hash 
   11     
   12     
   13     
   14     
   15     
   16     
   17     
   18     
   19     
   20     .dynstr

**Comment by jreiser** [comment]
@tklausmann UPX is not interested in such large executables.  If you are not interested in doing at least some custom programming, then using squashfs is the only available solution.

To increase communication and understanding, it would be helpful to know the reason(s) you are interested in compressing such a file, and the relative importance of each.  Is  is to reduce downloading time for software distribution, or to be more friendly for storage on typical consumer USB flash media, or what?

Be sure that the static linking step uses DT_RELR relocations, which in most cases dramatically reduces the storage that otherwise would be required for DT_RELA relocations.

Names of Segments and Sections that contain `fat` often are a hint that implementations for multiple environments have been concatenated.  Separate them earlier in the build+load pipeline.

Consider making multiple shared libraries from the several `LOAD` Segments that are just under 4MiB in size.

Perhaps the simplest extra programming is to make a shell script that de-compresses into a temporary file, then `exec` that file.

**Comment by tklausmann** [comment]
> To increase communication and understanding, it would be helpful to know the reason(s) you are interested in compressing such a file, and the relative importance of each. Is is to reduce downloading time for software distribution, or to be more friendly for storage on typical consumer USB flash media, or what?

I can of course only speak for what my team is doing. For us, the AMD/ROCm/HIP libraries (as shipped along with torch) are probably 90+% of the size of our workload docker images. Given the compression ratio I mentioned, we were wondering if we could reduce the size of the images (layers) at rest. usually, our workload only uses part of the code shipped, and even that is loaded once at startup and never changed. 

Ideally, upstream would provide shared libraries tailored to specific hardware models, i.e. you only need the .so's for your GPU model(s) and they're relatively small. I of course don't actually know the internal structure of the code, but my suspicion is that at least part of the huge size of these files is that they contain heavily-optimized code on a per-GPU/architecture basis.

The .so files we use are provided by upstream, and we only use them transitively --- our workload is a Python program that uses pytorch, which in turn uses the ROCm-specific versions of e.g. BLAS and similar libraries. AIUI, we _could_ build the binaries ourselves, but that would require acquiring and maintaining expertise in the build process and overall structure.

> Be sure that the static linking step uses DT_RELR relocations, which in most cases dramatically reduces the storage that otherwise would be required for DT_RELA relocations.
> 
> Names of Segments and Sections that contain `fat` often are a hint that implementations for multiple environments have been concatenated. Separate them earlier in the build+load pipeline.
> 
> Consider making multiple shared libraries from the several `LOAD` Segments that are just under 4MiB in size.
> 
> Perhaps the simplest extra programming is to make a shell script that de-compresses into a temporary file, then `exec` that file.

I appreciate your tips here (and I will definitely add them to my notes for when I have this kind of problem with my own compiled code), but since we don't really build the shlibs (see above), these approaches are not easy for us to implement. As I mentioned, ideally upstream would allow the consumers of these libraries to only ship what they need, but I also can understand their motivation to ship a "one size fits all" blob.
